packages/api/lambdas/cleanExecutions.js

```diff
@@ -1,48 +1,302 @@
+
+  const must = [
+    { range: { updatedAt: { lte: laterExpiration } } },
+    {
+      bool: {
+        should: [
+          { exists: { field: 'finalPayload' } },
+          { exists: { field: 'originalPayload' } },
+        ],
+      },
+    },
+  ];
+  const removePayloadScript = "ctx._source.remove('finalPayload'); ctx._source.remove('originalPayload')";
+  const mustNot = [];
+  let script = { inline: removePayloadScript };
+  if (runComplete && runNonComplete && (completeTimeoutDays !== nonCompleteTimeoutDays)) {
+    const removeForCompleteBoolean = `ctx._source.updatedAt < ${completeExpiration}L && ctx._source.status == 'completed'`;
+    const removeForNonCompleteBoolean = `ctx._source.updatedAt < ${nonCompleteExpiration}L && ctx._source.status != 'completed'`;
+    // a way to perform only integer comparison whenever possible
+    // and do relatively slow string comparison only when necessary
+    const removeForEitherBoolean = `ctx._source.updatedAt < ${earlierExpiration}L`;
+    const removeForLaterBoolean = (
+      completeExpiration === laterExpiration
+        ? removeForCompleteBoolean
+        : removeForNonCompleteBoolean
+    );
+    script = {
+      inline: `if ((${removeForEitherBoolean}) || (${removeForLaterBoolean})) { ${removePayloadScript} }`,
+    };
+  } else if (runNonComplete && !runComplete) {
+    mustNot.push({ term: { status: 'completed' } });
+  } else if (runComplete && !runNonComplete) {
+    must.push({ term: { status: 'completed' } });
+  }
+  const body = {
+    query: {
+      bool: {
+        must,
+        mustNot,
+      },
+    },
+    script: script,
+  };
+  const esClient = await getEsClient();
+  await esClient._client.updateByQuery({
+    index,
+    type: 'execution',
+    size: updateLimit,
+    body,
   });
+};
+
+/**
+ * Clean up PG executions that have expired
+ *
+ * @param {number} completeTimeoutDays - Maximum number of days a completed
+ *   record may have payload entries
+ * @param {number} nonCompleteTimeoutDays - Maximum number of days a non-completed
+ *   record may have payload entries
+ * @param {boolean} runComplete - Enable removal of completed execution
+ *   payloads
+ * @param {boolean} runNonComplete - Enable removal of execution payloads for
+ *   statuses other than 'completed'
+ * @returns {Promise<void>}
+*/
+const cleanupExpiredPGExecutionPayloads = async (
+  completeTimeoutDays,
+  nonCompleteTimeoutDays,
+  runComplete,
+  runNonComplete
+) => {
+  const {
+    laterExpiration,
+    completeExpiration,
+    nonCompleteExpiration,
+  } = getExpirationDates(
+    completeTimeoutDays,
+    nonCompleteTimeoutDays,
+    runComplete,
+    runNonComplete
+  );
+  const knex = await getKnexClient();
+  const updateLimit = Number(process.env.UPDATE_LIMIT || 10000);
+  const executionModel = new ExecutionPgModel();
+  const executionRecords = await getExpirablePayloadRecords(
+    knex,
+    laterExpiration,
+    updateLimit
+  );
```

Comment:
*if* they are not out of sync, they should remain in sync, although existing synchronicity issues will propagate. I wanted to keep them separate because successfully cleaning up everything seemed the priority. I dont' want to leave es records hanging around just because they aren't in pg for some reason.
on repeated runs the intention is that this gets *everything*

Reply:
What if a delete for the ES record fails? I think it would then have successfully deleted from PG and not from ES, thus causing a sync issue that didn't exist prior?

Reply:
I've more tightly synced these in a way that I think solves this problem

Reply:
I think this does make sense. `wipePayloads` will try to delete the ES payloads and (if successful) the PG payload. If the PG update fails, we still would have wiped the ES payload. I'm not sure it would be worth going to the trouble of reverting ES like we do in other updates. In the case that a PG update fails, we fail the promise. 

I think this makes sense. It may be worth bringing this up in a team setting to see if anyone disagrees with this approach. Being cautious because this will be an operation on potentially _many_ records.